### PR TITLE
Analyzer bug 61

### PR DESCRIPTION
**Pull request by ekcasey**
Doesn't write metadata written for `build = true` layers
Rewrites metadata for launch layers from then previous image label, even when the cached layer contents are valid

### CODE REVIEW

File: cache.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+		layersDir:   layersDir,
+		buildpackID: buildpackID,
+		layers:      layers,
+	}, nil
+}
+
+type cacheType int
+
+const (
+	staleNoMetadata cacheType = iota
```
Comment (sclevine) [inline cache.go]: I'd suggest prefixing these. Not sure we want a global constant called `valid`. Maybe:
```
layerValid = iota
layerStaleNoMetadata
layerStaleWrongSHA
layerNotForLaunch
```

File: testdata/analyzer/cached-layers/buildpack.node/node_modules.toml [hunk 2]
```diff
@@ -1,2 +1,3 @@
+launch = true
 [metadata]
   stalekey = "staleval"
```
Comment (sclevine) [inline testdata/analyzer/cached-layers/buildpack.node/node_modules.toml]: Missing newline

File: cache.go (hunk 1 above)
Comment (ekcasey) [reply cache.go]: I like the suggestion but how do you feel about prefixing with `cache` instead?

File: cache.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+)
+
+type cache struct {
+	layersDir   string
+	buildpackID string
+	layers      map[string]*LayerMetadata
+}
+
+func readCache(layersDir, buildpackID string) (cache, error) {
+	layers := make(map[string]*LayerMetadata)
```
Comment (sclevine) [inline cache.go]: Can use `layers := map[string]*LayerMetadata` if no pre-allocation is needed.

File: cache.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	tomls, err := filepath.Glob(filepath.Join(layersDir, buildpackID, "*.toml"))
+	if err != nil {
+		return cache{}, err
+	}
+	for _, toml := range tomls {
+		name := strings.TrimRight(filepath.Base(toml), ".toml")
+		layers[name], _ = readTOML(toml)
```
Comment (sclevine) [inline cache.go]: Seems like we should check this error?

File: cache.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+const (
+	staleNoMetadata cacheType = iota
+	staleWrongSHA
+	nonLaunch // we can't determine whether the cache is stale for launch=false layers
+	valid
+)
+
+func (c *cache) classifyLayer(layerName string, metadata *BuildpackMetadata) cacheType {
+	cachedLayer, ok := c.layers[layerName]
```
Comment (sclevine) [inline cache.go]: `ok` is defined but not used

File: analyzer_test.go [hunk 6]
```diff
@@ -176,9 +190,37 @@ func testAnalyzer(t *testing.T, when spec.G, it spec.S) {
...
+						} else {
+							expected := `build = false
+launch = true
+cache = false
+
+[metadata]
+  akey = "avalue"
+  bkey = "bvalue"
+`
+							actual := string(txt)
```
Comment (sclevine) [inline analyzer_test.go]: Could inline, not reused

File: analyzer.go [hunk 7]
```diff
@@ -49,179 +46,61 @@ func (a *Analyzer) analyze(metadata AppImageMetadata) error {
...
-	}
-	if err := os.Remove(abd.layerPath(name) + ".toml"); err != nil && !os.IsNotExist(err) {
-		return err
-	}
-	return nil
-}
-
-func appImageMetadata(groupBP string, metadata AppImageMetadata) (*BuildpackMetadata, bool) {
+func bpMetadata(buildpackID string, metadata AppImageMetadata) *BuildpackMetadata {
 	for _, buildpackMetaData := range metadata.Buildpacks {
```
Comment (sclevine) [inline analyzer.go]: Metadata is one word (`MetaData` -> `Metadata`)

File: cache.go (hunk 1 above)
Comment (sclevine) [reply cache.go]: Definitely like `cacheX` better!

**Pull request merged**